### ISSUE TITLE
Wrong data relocation optimization when symbol is hidden

### ISSUE THREAD

**Issue by bysui** [ClangBuiltLinux/linux#1844]
Hello,

**Background**
When running in an identity mapping address (low address), RIP-relative would retrieve the wrong symbol value. Therefore, I am attempting to use data relocation to obtain the correct symbol value in my [X86/PIE Patch](https://lore.kernel.org/lkml/bcdf9c31498fbd143848c15639d0c4fa46d804d6.1682673543.git.houwenlong.hwl@antgroup.com/). However, there are differences in behavior between GCC and CLANG.

GCC: 12.2.0
CLANG: 15.0.7

**Testcase**
So I write a testcase as following:
```C
#pragma GCC visibility push(hidden)
#include <stdio.h>

extern char _text[];

#define SYM_ABS_VAL(sym)       \
        ({ static unsigned long __##sym = (unsigned long)sym; __##sym; })

void test(void)
{
	unsigned long text_base = SYM_ABS_VAL(_text);

	printf("_text is %ld\n", text_base);
}
```

For GCC：

> gcc -O2 -c -fPIE test.c -o test.o
> objdump -r -S test.o

The output is:
```
test.o:     file format elf64-x86-64


Disassembly of section .text:

0000000000000000 <test>:
   0:	48 8b 35 00 00 00 00 	mov    0x0(%rip),%rsi        # 7 <test+0x7>
			3: R_X86_64_PC32	.data.rel.ro.local-0x4
   7:	48 8d 3d 00 00 00 00 	lea    0x0(%rip),%rdi        # e <test+0xe>
			a: R_X86_64_PC32	.LC0-0x4
   e:	31 c0                	xor    %eax,%eax
  10:	e9 00 00 00 00       	jmp    15 <test+0x15>
			11: R_X86_64_PLT32	printf-0x4
```

It could generate right data relocation.

For CLANG:
> clang -O2 -c -fPIE test.c -o test.o
> llvm-objdump -r -S test.o

The output is:
```
test.o:	file format elf64-x86-64

Disassembly of section .text:

0000000000000000 <test>:
       0: 48 8d 3d 00 00 00 00         	leaq	(%rip), %rdi            # 0x7 <test+0x7>
		0000000000000003:  R_X86_64_PC32	.L.str-0x4
       7: 48 8d 35 00 00 00 00         	leaq	(%rip), %rsi            # 0xe <test+0xe>
		000000000000000a:  R_X86_64_PC32	_text-0x4
       e: 31 c0                        	xorl	%eax, %eax
      10: e9 00 00 00 00               	jmp	0x15 <test+0x15>
		0000000000000011:  R_X86_64_PLT32	printf-0x4
```

It appears that CLANG directly omits data relocation. I am unsure if this behavior is intentional or not.

Thanks!

**Comment by MaskRay** [comment]
tl;dr Clang has the intended behavior.

(
"data relocation" is not a well-defined term.  The x86-64 psABI doesn't use "data relocation".
AArch64 uses "GOT-relative data relocations". I think "data" is mainly to state that this is not code, and the emphasis is not to say that there is a term "data relocation".
)

`#pragma GCC visibility push(hidden)` gives non-local declarations hidden visibility by default.
This applies to definitions like `test` and non-definition declarations like `extern char _text[];` and `printf`.

So GCC -fPIE somehow takes a detour to obtain the address of `_text`:

```
  movq    ___text.0(%rip), %rsi

.section        .data.rel.ro.local,"aw"
___text.0:
  .quad   _text
  .hidden _text
```

while Clang obtains the address directly `leaq    _text(%rip), %rsi`.
This is allowed by the small and kernel code models.
The code sequence is different but more efficient.

**Comment by bysui** [comment]
Thanks for elaborating.

I have switched to using movabsq to obtain the value of a symbol.
